evm/db/shard.py

```diff
@@ -73,6 +73,12 @@ def get_collation_by_hash(self, collation_hash: Hash32) -> Collation:
     # Canonical Collations
     #
     def set_canonical(self, header: CollationHeader) -> None:
+        try:
+            self.get_header_by_hash(header.hash)
+        except CollationHeaderNotFound:
+            raise ValueError("Cannot set unknown header as canonical: {}".format(
```

Comment:
Maybe re-raise `CollactionHeaderNotFound` with this message instead of using `ValueError`?